src/Database/LSMTree/Internal/WriteBufferReader.hs

```diff
@@ -0,0 +1,155 @@
+new :: forall m h.
+     (MonadMVar m, MonadST m, MonadMask m)
+  => TempRegistry m
+  -> HasFS m h
+  -> HasBlockIO m h
+  -> WriteBufferFsPaths
+  -> m (WriteBufferReader m h)
+new reg readerHasFS readerHasBlockIO fsPaths = do
+  (readerKOpsHandle :: FS.Handle h) <- FS.hOpen readerHasFS (Paths.writeBufferKOpsPath fsPaths) FS.ReadMode
+  -- Double the file readahead window (only applies to this file descriptor)
+  FS.hAdviseAll readerHasBlockIO readerKOpsHandle FS.AdviceSequential
+  readerWriteBufferBlobs <- allocateTemp reg (WBB.open readerHasFS (Paths.writeBufferBlobPath fsPaths) FS.AllowExisting) releaseRef
+  -- Load first page from disk, if it exists.
+  readerCurrentEntryNo <- newPrimVar (0 :: Word16)
+  firstPage <- readDiskPage readerHasFS readerKOpsHandle
+  readerCurrentPage <- newMutVar firstPage
+  pure $ WriteBufferReader{..}
+
+{-# SPECIALISE
+  next ::
+       WriteBufferReader IO h
+    -> IO (Result IO h)
+  #-}
+-- | See 'Database.LSMTree.Internal.RunReader.next'.
+next :: forall m h.
+     (MonadSTM m, MonadST m, MonadMask m)
+  => WriteBufferReader m h
+  -> m (Result m h)
+next WriteBufferReader {..} = do
+    readMutVar readerCurrentPage >>= \case
+      Nothing ->
+        return Empty
+      Just page -> do
+        entryNo <- readPrimVar readerCurrentEntryNo
+        go entryNo page
+  where
+    go :: Word16 -> RawPage -> m (Result m h)
+    go !entryNo !page =
+        -- take entry from current page (resolve blob if necessary)
+        case rawPageIndex page entryNo of
+          IndexNotPresent -> do
+            -- if it is past the last one, load a new page from disk, try again
+            newPage <- readDiskPage readerHasFS readerKOpsHandle
+            stToIO $ writeMutVar readerCurrentPage newPage
+            case newPage of
+              Nothing -> do
+                return Empty
+              Just p -> do
+                writePrimVar readerCurrentEntryNo 0
+                go 0 p  -- try again on the new page
+          IndexEntry key entry -> do
+            modifyPrimVar readerCurrentEntryNo (+1)
+            let entry' :: E.Entry SerialisedValue (RawBlobRef m h)
+                entry' = fmap (WBB.mkRawBlobRef readerWriteBufferBlobs) entry
+            let rawEntry = Entry entry'
+            return (ReadEntry key rawEntry)
+          IndexEntryOverflow key entry lenSuffix -> do
+            -- TODO: we know that we need the next page, could already load?
+            modifyPrimVar readerCurrentEntryNo (+1)
+            let entry' :: E.Entry SerialisedValue (RawBlobRef m h)
+                entry' = fmap (WBB.mkRawBlobRef readerWriteBufferBlobs) entry
+            overflowPages <- readOverflowPages readerHasFS readerKOpsHandle lenSuffix
+            let rawEntry = mkEntryOverflow entry' page lenSuffix overflowPages
+            return (ReadEntry key rawEntry)
+
+close :: WriteBufferReader m h -> m ()
+close WriteBufferReader{..} = FS.hClose readerHasFS readerKOpsHandle
```

Comment:
`close` should also release the `readerWriteBufferBlobs` reference, if we make the changes I suggested above